File: app/src/main/java/com/sakuna63/tumblrin/data/repository/PostRepository.java
package com.sakuna63.tumblrin.data.repository;

import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import com.sakuna63.tumblrin.application.di.scope.ActivityScope;
import com.sakuna63.tumblrin.data.dao.PostDao;
import com.sakuna63.tumblrin.data.datasource.PostDataSource;
import com.sakuna63.tumblrin.data.model.Post;
import com.sakuna63.tumblrin.data.model.response.PostsResponse;
import com.sakuna63.tumblrin.data.model.response.TumblrResponse;
import com.sakuna63.tumblrin.data.net.TumblrService;

import java.util.List;

import javax.inject.Inject;

import rx.Single;
import rx.functions.Func1;

@ActivityScope
public class PostRepository implements PostDataSource {

    @NonNull
    private final TumblrService service;

    @NonNull
    private final PostDao postDao;

    @Inject
    public PostRepository(@NonNull TumblrService service,
                          @NonNull PostDao postDao) {
        this.service = service;
        this.postDao = postDao;
    }

    @Override
    public Single<List<Post>> getDashboard(@IntRange(from = 1, to = 20) Integer limit,
                                           @IntRange(from = 0) Integer offset, Long sinceId) {
        return service.getDashboard(limit, offset, sinceId, false, false)
                .map(new Func1<TumblrResponse<PostsResponse>, List<Post>>() {
                    @Override
                    public List<Post> call(final TumblrResponse<PostsResponse> response) {
                        List<Post> posts = response.getResponse().getPosts();
                        postDao.insert(posts);
                        return posts;
                    }
                });
    }
}

File: app/src/main/java/com/sakuna63/tumbin/data/repository/PostRepository.kt
package com.sakuna63.tumbin.data.repository

import android.support.annotation.IntRange
import com.sakuna63.tumbin.data.api.TumblrService
import com.sakuna63.tumbin.data.dao.DashboardRealmDao
import com.sakuna63.tumbin.data.datasource.PostDataSource
import rx.Completable
import rx.Single
import javax.inject.Inject

class PostRepository
@Inject
constructor(private val service: TumblrService,
            private val dashboardRealmDao: DashboardRealmDao) : PostDataSource {

    override fun fetchDashboard(@IntRange(from = 1, to = 20) limit: Int?,
                                @IntRange(from = 0) offset: Int?,
                                sinceId: Long?,
                                beforeId: Long?): Single<Int> {

        return service.getDashboard(limit, offset, sinceId, beforeId, true, true).map { response ->
            val posts = response.response.posts
            dashboardRealmDao.insert(posts)
            posts.size
        }
    }

    override fun deleteAllDashboardCache(): Completable =
            Completable.fromAction { dashboardRealmDao.deleteAll() }
}
